File: src/providers/supabase/rules/order-by-timestamp-not-identity.ts
/**
 * supabase-order-by-timestamp-not-identity (correctness)
 *
 * `.order("id", ...)` "works" only because a `bigint identity always` PK
 * happens to be monotonic with insert order today — a correctness smell
 * that breaks under any future bulk-insert/backfill/replication scenario.
 * When the same query already selects a purpose-built timestamp column
 * (e.g. `created_at`), that column — not the surrogate key — should drive
 * the ordering.
 *
 * Like scope-queries-by-tenant-column, this walks each query chain
 * bottom-up (`CallExpression:exit`) so the `.select()` that introduced the
 * candidate timestamp column(s) is already recorded by the time the
 * `.order()` built on top of it is visited.
 */
import { chainObjectCall, isTimestampColumnName, memberPropName, parseSelectColumns } from '../utils.js';

interface ChainState {
  timestampColumns: string[];
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Order by a selected timestamp column instead of the identity column',
      category: 'correctness',
      rationale:
        'Ordering by a bigint identity PK only produces correct chronological order because the PK happens to be monotonic with insert order today. That assumption breaks under bulk inserts, backfills, or replication where PK order and time order can diverge. When the query already selects a timestamp column built for this purpose, order by it instead of the surrogate key.',
      docsUrl: 'https://supabase.com/docs/reference/javascript/order',
      recommended: true,
    },
    messages: {
      orderByIdentity:
        'This query selects "{{column}}" but orders by "id" instead. Order by "{{column}}" so result order does not depend on PK/insert-order coincidence.',
    },
    schema: [],
  },
  create(context: any) {
    const chainStates = new Map<any, ChainState>();

    return {
      'CallExpression:exit'(node: any) {
        const prop = memberPropName(node);
        if (!prop) return;

        const objCall = chainObjectCall(node);

        if (prop === 'select' && objCall && memberPropName(objCall) === 'from') {
          const columns = parseSelectColumns(node.arguments?.[0]);
          const timestampColumns = columns.filter(isTimestampColumnName);
          chainStates.set(node, { timestampColumns });
          return;
        }

        const state = objCall ? chainStates.get(objCall) : undefined;
        if (!state) return;

        if (prop === 'order') {
          const colArg = node.arguments?.[0];
          const orderColumn = colArg?.type === 'Literal' ? colArg.value : undefined;
          if (
            typeof orderColumn === 'string' &&
            orderColumn.toLowerCase() === 'id' &&
            state.timestampColumns.length > 0
          ) {
            context.report({
              node,
              messageId: 'orderByIdentity',
              data: { column: state.timestampColumns[0] },
            });
          }
        }

        chainStates.set(node, state);
      },
    };
  },
};

export const supabaseOrderByTimestampNotIdentityRule = rule;
export default rule;

File: src/providers/supabase/rules/python/order_by_timestamp_not_identity.py
"""supabase-order-by-timestamp-not-identity

Parity with the JS rule: `.order("id", ...)` "works" only because a bigint
identity PK happens to be monotonic with insert order today. When the same
query already selects a purpose-built timestamp column (e.g. `created_at`),
order by that column instead of the surrogate key.

Same bottom-up chain walk as scope-queries-by-tenant-column.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (
    call_report_loc,
    chain_object_call,
    is_supabase_table_call,
    is_timestamp_column_name,
    iter_calls_post_order,
    member_prop_name,
    parse_select_columns,
)

RULE_KEY = "supabase-order-by-timestamp-not-identity"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    chain_states: dict[int, dict] = {}

    for node in iter_calls_post_order(tree):
        prop = member_prop_name(node)
        if not prop:
            continue

        obj_call = chain_object_call(node)

        if prop == "select" and obj_call is not None and is_supabase_table_call(obj_call):
            columns = parse_select_columns(node.args[0] if node.args else None)
            timestamp_columns = [c for c in columns if is_timestamp_column_name(c)]
            chain_states[id(node)] = {"timestamp_columns": timestamp_columns}
            continue

        state = chain_states.get(id(obj_call)) if obj_call is not None else None
        if state is None:
            continue

        if prop == "order":
            col_arg = node.args[0] if node.args else None
            order_column = col_arg.value if isinstance(col_arg, ast.Constant) else None
            if (
                isinstance(order_column, str)
                and order_column.lower() == "id"
                and state["timestamp_columns"]
            ):
                line, col = call_report_loc(node)
                out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

        chain_states[id(node)] = state

    return out
